x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/ValueSources.java

```diff
@@ -46,18 +47,22 @@ public static List<ValueSourceInfo> sources(
                 continue; // the field does not exist in this context
             }
 
-            // MatchOnlyTextFieldMapper class lives in the mapper-extras module. We use string equality
-            // for the field type name to avoid adding a dependency to the module
-            if (fieldType instanceof TextFieldMapper.TextFieldType || "match_only_text".equals(fieldType.typeName())) {
-                ValuesSource vs = textValueSource(ctx, fieldType);
-                sources.add(new ValueSourceInfo(CoreValuesSourceType.KEYWORD, vs, elementType, ctx.getIndexReader()));
-                continue;
-            }
+            if (fieldType.hasDocValues() == false) {
+                // MatchOnlyTextFieldMapper class lives in the mapper-extras module. We use string equality
+                // for the field type name to avoid adding a dependency to the module
+                if (fieldType instanceof KeywordFieldMapper.KeywordFieldType
+                    || fieldType instanceof TextFieldMapper.TextFieldType
+                    || "match_only_text".equals(fieldType.typeName())) {
```

Comment:
Could be extracted into a constant.

Reply:
I'd left a comment on the other copy of this about how it'd be lovely if this could go backwards, with the mappers defining this behavior. But that's not worth the extra complexity now.

Reply:
I have thought about it too. Probably, the least invasive way to implement it in the field mappers is to add a new `MappedFieldType.FielddataOperation.ESQL` and use that in the `MappedFieldType.fielddataBuilder()` to define the behavior we need. I can give  it a try in a following PR